In [1]:
import tensorflow as tf

### 차원에 따른 텐서 생성

In [2]:
t0 = tf.constant(1)
print(t0)
print(tf.rank(t0))  # 축의 개수를 반환합니다.

tf.Tensor(1, shape=(), dtype=int32)
tf.Tensor(0, shape=(), dtype=int32)


In [3]:
t1 = tf.constant([1, 2, 3])
print(t1)
print(tf.rank(t1))

tf.Tensor([1 2 3], shape=(3,), dtype=int32)
tf.Tensor(1, shape=(), dtype=int32)


In [4]:
t2 = tf.constant([[1, 2, 3], [1, 2, 3]])
print(t2)
print(tf.rank(t2))

tf.Tensor(
[[1 2 3]
 [1 2 3]], shape=(2, 3), dtype=int32)
tf.Tensor(2, shape=(), dtype=int32)


In [5]:
t3 = tf.constant([[[1, 2, 3], [1, 2, 3], [1, 2, 3]]])
print(t3)
print(tf.rank(t3))

tf.Tensor(
[[[1 2 3]
  [1 2 3]
  [1 2 3]]], shape=(1, 3, 3), dtype=int32)
tf.Tensor(3, shape=(), dtype=int32)


### 텐서의 타입

In [6]:
# 상수값 2인 텐서 생성
i = tf.constant(2)
print(i)

tf.Tensor(2, shape=(), dtype=int32)


* 타입 int32

In [7]:
# 상수값 2.0인 텐서생성
f = tf.constant(2.0)
print(f)

tf.Tensor(2.0, shape=(), dtype=float32)


In [8]:
s = tf.constant("Whathe")
print(s)

tf.Tensor(b'Whathe', shape=(), dtype=string)


In [9]:
# 16비트 실수형으로 직접 타입 지정하기
f16 = tf.constant(2., dtype=tf.float16)
print(f16)

tf.Tensor(2.0, shape=(), dtype=float16)


In [18]:
# 8비트 정수형으로 타입 지정
t8 = tf.constant(3, dtype=tf.int8)
print(t8)

tf.Tensor(3, shape=(), dtype=int8)


### 텐서 타입 변환

In [17]:
# 16비트 실수형을 32비트 실수형으로 변환
f32 = tf.cast(f16, tf.float32)
print(f32)
print(f16)  # 새로운 변수에 지정해서 대상은 바뀌지 않음

tf.Tensor(2.0, shape=(), dtype=float32)
tf.Tensor(2.0, shape=(), dtype=float16)


In [20]:
# 8비트 정수형에서 32비트 정수형으로 변환
t32 = tf.cast(t8, tf.int32)
print(t32)
print(t8)

tf.Tensor(3, shape=(), dtype=int32)
tf.Tensor(3, shape=(), dtype=int8)


### 텐서 형상 변환

In [23]:
# reshape
x = tf.constant([[1], [2], [3]])
print(x)
print(x.shape)

y = tf.reshape(x, [1, 3])
print(y)
print(y.shape)

tf.Tensor(
[[1]
 [2]
 [3]], shape=(3, 1), dtype=int32)
(3, 1)
tf.Tensor([[1 2 3]], shape=(1, 3), dtype=int32)
(1, 3)


In [24]:
# 텐서 전치(transpose)
print(y)
print(tf.transpose(y))
print(y.shape)

tf.Tensor([[1 2 3]], shape=(1, 3), dtype=int32)
tf.Tensor(
[[1]
 [2]
 [3]], shape=(3, 1), dtype=int32)
(1, 3)


### 차원 압축

In [26]:
# tf.squeeze 는 텐서에서 크기가 1인 차원을 제거함
print(x)
print(tf.squeeze(x))

tf.Tensor(
[[1]
 [2]
 [3]], shape=(3, 1), dtype=int32)
tf.Tensor([1 2 3], shape=(3,), dtype=int32)


### 차원 추가

In [27]:
# tf.expand_dims는 차원을 추가함, axis는 차원을 확장할 텐서의 축을 지정함
print(y)
print(tf.expand_dims(y, axis=0))
print(tf.expand_dims(y, axis=1))
print(tf.expand_dims(y, axis=2))

tf.Tensor([[1 2 3]], shape=(1, 3), dtype=int32)
tf.Tensor([[[1 2 3]]], shape=(1, 1, 3), dtype=int32)
tf.Tensor([[[1 2 3]]], shape=(1, 1, 3), dtype=int32)
tf.Tensor(
[[[1]
  [2]
  [3]]], shape=(1, 3, 1), dtype=int32)


### 텐서 분리

In [29]:
# tf.split 로 차원을 지정하여 분리 아래에서는 텐서 x를 3개로 분리
print(x)
print(tf.split(x, 3))

tf.Tensor(
[[1]
 [2]
 [3]], shape=(3, 1), dtype=int32)
[<tf.Tensor: shape=(1, 1), dtype=int32, numpy=array([[1]])>, <tf.Tensor: shape=(1, 1), dtype=int32, numpy=array([[2]])>, <tf.Tensor: shape=(1, 1), dtype=int32, numpy=array([[3]])>]


### 텐서 연결

In [30]:
# tf.concat 은 axis로 지정한 축을 기준으로 텐서들을 이어붙임
print(x)
print(tf.concat([x, x], axis=0))
print(tf.concat([x, x], axis=1))

tf.Tensor(
[[1]
 [2]
 [3]], shape=(3, 1), dtype=int32)
tf.Tensor(
[[1]
 [2]
 [3]
 [1]
 [2]
 [3]], shape=(6, 1), dtype=int32)
tf.Tensor(
[[1 1]
 [2 2]
 [3 3]], shape=(3, 2), dtype=int32)


### 텐서 연산

#### 0차원 텐서 연산

In [31]:
# 1,2줄과 3,4줄은 같은것
print(tf.constant(2) + tf.constant(2))
print(tf.constant(2) - tf.constant(2))
print(tf.add(tf.constant(2), tf.constant(2)))
print(tf.subtract(tf.constant(2), tf.constant(2)))

tf.Tensor(4, shape=(), dtype=int32)
tf.Tensor(0, shape=(), dtype=int32)
tf.Tensor(4, shape=(), dtype=int32)
tf.Tensor(0, shape=(), dtype=int32)


In [32]:
print(tf.constant(2) * tf.constant(2))
print(tf.constant(2) / tf.constant(2))
print(tf.multiply(tf.constant(2), tf.constant(2)))
print(tf.divide(tf.constant(2), tf.constant(2)))

tf.Tensor(4, shape=(), dtype=int32)
tf.Tensor(1.0, shape=(), dtype=float64)
tf.Tensor(4, shape=(), dtype=int32)
tf.Tensor(1.0, shape=(), dtype=float64)


In [35]:
# 다른타입의 연산들은 변환 해주면서 연산해야함
print(tf.cast(tf.constant(2), tf.float32) + tf.constant(2.2))

tf.Tensor(4.2, shape=(), dtype=float32)


#### 1차원 이상의 텐서 연산

In [38]:
t1 = tf.constant([1, 2, 3])
t2 = tf.constant([[1, 2, 3], [1, 2, 3]])
print(t1)
print(t2)

tf.Tensor([1 2 3], shape=(3,), dtype=int32)
tf.Tensor(
[[1 2 3]
 [1 2 3]], shape=(2, 3), dtype=int32)


In [39]:
print(t1 + t1)
print(t1 + t2)

tf.Tensor([2 4 6], shape=(3,), dtype=int32)
tf.Tensor(
[[2 4 6]
 [2 4 6]], shape=(2, 3), dtype=int32)


In [37]:
print(t2 + t2)
print(t2 - t2)
print(t2 * t2)
print(t2 / t2)

tf.Tensor(
[[2 4 6]
 [2 4 6]], shape=(2, 3), dtype=int32)
tf.Tensor(
[[0 0 0]
 [0 0 0]], shape=(2, 3), dtype=int32)
tf.Tensor(
[[1 4 9]
 [1 4 9]], shape=(2, 3), dtype=int32)
tf.Tensor(
[[1. 1. 1.]
 [1. 1. 1.]], shape=(2, 3), dtype=float64)


In [40]:
c = tf.constant([[4.0, 5.0, 6.0], 
                 [10.0, 9.0, 8.0]])

print(tf.reduce_max(c))
print(tf.argmax(c))
print(tf.nn.softmax(c))

tf.Tensor(10.0, shape=(), dtype=float32)
tf.Tensor([1 1 1], shape=(3,), dtype=int64)
tf.Tensor(
[[0.09003057 0.24472848 0.66524094]
 [0.66524094 0.24472848 0.09003057]], shape=(2, 3), dtype=float32)


* reduce_max(): 텐서 값 중에서 최대값을 계산합니다.
* argmax(): 최대값의 위치를 반환합니다. 예제에서는 [4.0, 5.0, 6.0] 보다 [10.0, 9.0, 8.0]이 다 크기 때문에 [0 0 0]이 아니라 [1 1 1]의 결과를 보여주고 있습니다.
* nn.softmax(): 텐서의 값을 0과 1 사이의 값으로 보여줍니다. 예제에서는 [4.0, 5.0, 6.0] 값이 [0.09003057 0.24472848 0.66524094]으로 변환되었고, [10.0, 9.0, 8.0]이 [0.66524094 0.24472848 0.09003057]로 변환되었습니다.